klippy/extras/sht4x.py

```diff
@@ -0,0 +1,262 @@
+        self.printer = config.get_printer()
+        self.name = config.get_name().split()[-1]
+        self.reactor = self.printer.get_reactor()
+
+        # I2C setup
+        valid_addresses = [0x44, 0x45, 0x46]  # From datasheet
+        addr = config.getint('i2c_address', DEFAULT_ADDR)
+        if addr not in valid_addresses:
+            addr_list = ', '.join(['0x%x' % a for a in valid_addresses])
+            raise config.error("Invalid I2C address 0x%x. "
+                              "Valid addresses are: %s" % (addr, addr_list))
+
+        speed = config.getint('i2c_speed', 100000)
+        self.i2c = bus.MCU_I2C_from_config(
+            config, default_addr=addr, default_speed=speed)
+        self.mcu = self.i2c.get_mcu()
+
+        # Precision setup
+        precision = config.get('precision', 'high').lower()
+        precision_map = {
+            'high': HIGH_PRECISION_MODE,
+            'medium': MEDIUM_PRECISION_MODE,
+            'low': LOW_PRECISION_MODE
+        }
+        if precision not in precision_map:
+            raise config.error("Invalid precision value '%s'. "
+                              "Valid options are: high, medium, low"
+                              % precision)
+        self.precision_mode = precision_map[precision]
+
+        # Core sensor state
+        self.temp = 0
+        self.humidity = 0
+        self._error_count = 0
+        self.sensor_ready = False
+        self.measurement_lock = threading.Lock()
+        self._callback = None
+        self.serial_number = None
+
+        # Register sensor
+        self.printer.add_object("sht4x " + self.name, self)
+        self.printer.register_event_handler("klippy:connect",
+                                           self.handle_connect)
+
+    def handle_connect(self):
+        """Initialize sensor"""
+        try:
+            self.reset()
+            # Add a small delay after reset
+            self.reactor.pause(self.reactor.monotonic() + 0.01)  # 10ms
+
+            # Try to read serial number
+            self.serial_number = self.get_serial_number()
+            if self.serial_number:
+                logging.info("SHT4X: Sensor serial number: 0x%08X",
+                           self.serial_number)
+
+            # Test measurement
+            if self._test_measurement():
+                self.sensor_ready = True
+                logging.info("SHT4X: Sensor ready")
+                self.sample_timer = self.reactor.register_timer(
+                    self.sample_sensor)
+                self.reactor.update_timer(self.sample_timer,
+                                        self.reactor.NOW)
+            else:
+                logging.error("SHT4X: Sensor test failed")
+        except Exception as e:
+            logging.error("SHT4X: Init failed: %s" % str(e))
+
+    def _test_measurement(self):
+        """Quick measurement test"""
+        try:
+            recv = self.get_measurements()
+            if len(recv) != 6:
```

Comment:
You can't receive less or more than what you have been requested.
If you somehow get this, this is probably a serious bug.